File: lib/to_mdx/parsers/open_document/draw/group.rb
# frozen_string_literal: true

require_relative "../base"
require_relative "../mixins/positionable_element"

module ToMdx
  module Parsers
    module Draw
      class Group < Base
        include Mixins::PositionableElement

        def initialize(element = nil, context = {})
          super(element, context)
          @draw_style_name = context[:styles].use(element&.attributes&.[]("draw:style-name"))
        end

        def to_mdx
          draw_style_class = @draw_style_name.empty? ? "" : " className=\"#{@draw_style_name}\""
          <<~MDX
            <div#{draw_style_class} data-name="#{@element.attributes["draw:name"]}">
              #{contentful_children.map(&:to_mdx).compact.join}
            </div>
          MDX
        end

        def to_s
          contentful_children.map(&:to_s).compact.join
        end

        register_for "draw:g"
      end
    end
  end
end

File: src/parsers/open-document/draw/group.ts
import dedent from "dedent-js";
import type { Style } from "../../../styles.ts";
import type { ParseContext } from "../../base_element.ts";
import { BaseElement } from "../../base_element.ts";

export class Group extends BaseElement {
  private name: string;
  private drawStyleName: Style;

  constructor(element: Element, context: ParseContext, parent: BaseElement | null) {
    super(element, context, parent);
    this.name = this.attr("draw:name") || "";
    this.drawStyleName = context.styles.use(this.attr("draw:style-name"));
  }

  toMdx(): string {
    const drawStyleClass = this.drawStyleName.empty() ? "" : ` className="${this.drawStyleName}"`;
    const content = this.contentfulChildren()
      .map((child) => child.toMdx())
      .filter((content) => content.trim().length > 0)
      .join("");

    return dedent`
      <div${drawStyleClass} data-name="${this.name}">
        ${content}
      </div>
    `;
  }

  toString(): string {
    return this.contentfulChildren()
      .map((child) => child.toString())
      .filter((content) => content.trim().length > 0)
      .join("");
  }

  static {
    BaseElement.registerFor.call(this, "draw:g");
  }
}
